ff/build.rs

```diff
@@ -1,17 +1,24 @@
-extern crate rustc_version;
-use rustc_version::{version_meta, Channel};
+use rustc_version::Version;
 
 fn main() {
     println!("cargo:rerun-if-changed=build.rs");
 
-    let is_nightly = version_meta().expect("nightly check failed").channel == Channel::Nightly;
+    let target_arch = std::env::var("CARGO_CFG_TARGET_ARCH").unwrap();
 
-    let should_use_asm = cfg!(all(
-        feature = "asm",
-        target_feature = "bmi2",
-        target_feature = "adx",
-        target_arch = "x86_64"
-    )) && is_nightly;
+    let rust_version = rustc_version::version().unwrap();
+    let is_asm_stable = rust_version >= Version::new(1, 59, 0);
+
+    let asm_enabled = cfg!(feature = "asm");
+
+    let is_x86_64 = target_arch == "x86_64";
+    let mut bmi2_enabled = false;
+    let mut adx_enabled = false;
+    if is_x86_64 {
+        bmi2_enabled = is_x86_feature_detected!("bmi2");
+        adx_enabled = is_x86_feature_detected!("adx");
+    }
+
+    let should_use_asm = is_asm_stable && asm_enabled && bmi2_enabled && adx_enabled && is_x86_64;
```

Comment:
Hm should we just get rid of this, and instead change our minimum supported rust version to 1.59? And in the places that we use the assembly backend, we can directly gate it with a 
```rust
#[cfg(
    feature = "asm",
    target_feature = "bmi2",
    target_feature = "adx",
    target_arch = "x86_64"
)]
```

Reply:
I think there is only one question: do we want to bump up the minimal supported rust version to 1.59?

That would clearly be ideal, and we can cut the `build.rs`.

Reply:
I think it's fine personally

Reply:
I would recommend doing that in a separate PR---the current stable is 1.58.1, and it would need 20+ days to be 1.59.0.

So, if we drop it, for the next three weeks, we may have some trouble.

Reply:
I can add an issue and a separate PR.

Reply:
Hm then I would still move most of the `cfg` logic out of the `build.rs`, but emit the `asm` cfg in `build.rs`, so that the `cfg` gate becomes
```rust
#[cfg(all(
    asm,
    target_feature = "bmi2",
    target_feature = "adx",
    target_arch = "x86_64"
))]
```

Reply:
and the `build.rs` becomes the following? So it would in a shape that is ready to drop.

```
use rustc_version::Version;

fn main() {
    println!("cargo:rerun-if-changed=build.rs");
    
    let rust_version = rustc_version::version().unwrap();
    let is_asm_stable = rust_version >= Version::new(1, 59, 0);
    let asm_enabled = cfg!(feature = "asm");

    let should_use_asm = is_asm_stable && asm_enabled;
    if should_use_asm {
        println!("cargo:rustc-cfg=use_asm");
    }
}
```

Reply:
Yup, exactly!

Reply:
There is a complication: what do you think would be the best solution here?

https://github.com/arkworks-rs/algebra/blob/migrate-to-asm/ff/src/fields/models/fp/montgomery_backend.rs#L86

Reply:
ah, it seems that I can simply do:

```
if N <= 6
                && N > 1
                && cfg!(all(
                    asm,
                    target_feature = "bmi2",
                    target_feature = "adx",
                    target_arch = "x86_64"
                ))
            {
```

Reply:
We write `cfg!(all(
    use_asm,
    target_feature = "bmi2",
    target_feature = "adx",
    target_arch = "x86_64"
))`

Reply:
another complication---actually this is a mistake:
https://github.com/arkworks-rs/algebra/blob/migrate-to-asm/ff/src/biginteger/mod.rs#L242

It just detects the `asm` feature flag but not other things. 

Seems to be complicated. 

Should we do `cfg_attr!` in the `lib.rs`?

Reply:
That one is fine, as it doesn't require 1.59, nor `bmi2` or `adx`

Reply:
Changed the logic. It appears that rust version has to be queried in build.rs. The rest can be moved out. 

I use `inline_asm_stable` since not all asm requires 1.59.0, as you pointed out.

Reply:
`cfg(version(1.59))` is not yet stable, though, we can drop it when we decided that the minimal supported rust version is >=1.59.